source/loader/layers/sanitizer/asan_interceptor.cpp

```diff
@@ -0,0 +1,617 @@
+    auto Context = getContext(Queue);
+    auto Device = getDevice(Queue);
+    assert(Device != nullptr);
+
+    auto ContextInfo = getContextInfo(Context);
+
+    auto HostInfo = ContextInfo->getDeviceInfo(nullptr);
+    auto DeviceInfo = ContextInfo->getDeviceInfo(Device);
+    auto QueueInfo = ContextInfo->getQueueInfo(Queue);
+
+    std::shared_lock<ur_shared_mutex> HostGuard(HostInfo->Mutex,
+                                                std::defer_lock);
+    std::unique_lock<ur_shared_mutex> DeviceGuard(DeviceInfo->Mutex,
+                                                  std::defer_lock);
+    std::scoped_lock<std::shared_lock<ur_shared_mutex>,
+                     std::unique_lock<ur_shared_mutex>, ur_mutex>
+        Guard(HostGuard, DeviceGuard, QueueInfo->Mutex);
+
+    ur_event_handle_t LastEvent = QueueInfo->LastEvent;
+
+    // FIXME: Always update host USM, but it'd be better to update host USM
+    // selectively, or each devices once
+    for (auto &AllocInfo : HostInfo->AllocInfos) {
+        UR_CALL(enqueueAllocInfo(Context, Device, Queue, AllocInfo, LastEvent));
+    }
+
+    for (auto &AllocInfo : DeviceInfo->AllocInfos) {
+        UR_CALL(enqueueAllocInfo(Context, Device, Queue, AllocInfo, LastEvent));
+    }
+    DeviceInfo->AllocInfos.clear();
+
+    QueueInfo->LastEvent = LastEvent;
+
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t SanitizerInterceptor::addContext(ur_context_handle_t Context) {
+    auto ContextInfo = std::make_shared<ur_sanitizer_layer::ContextInfo>();
+
+    // Host Device
+    auto DeviceInfo = std::make_shared<ur_sanitizer_layer::DeviceInfo>();
+    DeviceInfo->Type = DeviceType::CPU;
+    DeviceInfo->Alignment = ASAN_SHADOW_GRANULARITY;
+
+    // TODO: Check if host asan is enabled
+    DeviceInfo->ShadowOffset = 0;
+    DeviceInfo->ShadowOffsetEnd = 0;
+
+    ContextInfo->DeviceMap.emplace(nullptr, std::move(DeviceInfo));
+
+    std::scoped_lock<ur_shared_mutex> Guard(m_ContextMapMutex);
+    m_ContextMap.emplace(Context, std::move(ContextInfo));
+
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t SanitizerInterceptor::removeContext(ur_context_handle_t Context) {
+    std::scoped_lock<ur_shared_mutex> Guard(m_ContextMapMutex);
+    assert(m_ContextMap.find(Context) != m_ContextMap.end());
+    m_ContextMap.erase(Context);
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t SanitizerInterceptor::addDevice(ur_context_handle_t Context,
+                                            ur_device_handle_t Device) {
+    auto DeviceInfo = std::make_shared<ur_sanitizer_layer::DeviceInfo>();
+
+    // Query device type
+    ur_device_type_t DeviceType;
+    UR_CALL(context.urDdiTable.Device.pfnGetInfo(
+        Device, UR_DEVICE_INFO_TYPE, sizeof(DeviceType), &DeviceType, nullptr));
+    switch (DeviceType) {
+    case UR_DEVICE_TYPE_CPU:
+        DeviceInfo->Type = DeviceType::CPU;
+        break;
+    case UR_DEVICE_TYPE_GPU:
+        DeviceInfo->Type = DeviceType::GPU_PVC;
```

Comment:
GPU_PVC is better to be a Name rather than a Type. Perhaps a device name entry is needed in DeviceInfo?

Reply:
Save it to a string will make the code a little slow, and I don't think using enum value as device type is a big problem.
I'll use the device name to check the device type in the future.

Reply:
Being a name doesn't need to be a string. You can define a new enum for the device name.
It is just that the above line doesn't make much sense because the left is device type and the right is device name.

Reply:
Makes sense. I'll change it.

Reply:
> You can define a new enum for the device name.

It seems it's hard to define the enum values of device name.